dodiscover/cd/kernel_test.py

```diff
@@ -0,0 +1,249 @@
+        e_hat = _estimate_propensity_scores(
+            K,
+            group_ind,
+            penalty=self.propensity_penalty_,
+            n_jobs=self.n_jobs,
+            random_state=self.random_state,
+        )
+
+        # now compute null distribution
+        null_dist = self.compute_null(
+            e_hat, K, L, null_reps=self.null_reps, random_state=self.random_state
+        )
+        self.null_dist_ = null_dist
+
+        # compute the pvalue
+        pvalue = (1 + np.sum(null_dist >= stat)) / (1 + self.null_reps)
+        return stat, pvalue
+
+    def _statistic(self, K: ArrayLike, L: ArrayLike, group_ind: ArrayLike) -> float:
+        n_samples = len(K)
+
+        # compute W matrices from K and z
+        W0, W1 = self._compute_inverse_kernel(K, group_ind)
+
+        # compute L kernels
+        first_mask = np.array(1 - group_ind, dtype=bool)
+        second_mask = np.array(group_ind, dtype=bool)
+        L0 = L[np.ix_(first_mask, first_mask)]
+        L1 = L[np.ix_(second_mask, second_mask)]
+        L01 = L[np.ix_(first_mask, second_mask)]
+
+        # compute the final test statistic
+        K0 = K[:, first_mask]
+        K1 = K[:, second_mask]
+        KW0 = K0 @ W0
+        KW1 = K1 @ W1
+
+        # compute the three terms in Lemma 4.4
+        first_term = np.trace(KW0.T @ KW0 @ L0)
+        second_term = np.trace(KW1.T @ KW0 @ L01)
+        third_term = np.trace(KW1.T @ KW1 @ L1)
+
+        # compute final statistic
+        stat = (first_term - 2 * second_term + third_term) / n_samples
+        return stat
+
+    def _compute_inverse_kernel(self, K, z) -> Tuple[ArrayLike, ArrayLike]:
+        """Compute W matrices as done in KCD test.
+
+        Parameters
+        ----------
+        K : ArrayLike of shape (n_samples, n_samples)
+            The kernel matrix.
+        z : ArrayLike of shape (n_samples)
+            The indicator variable of 1's and 0's for which samples belong
+            to which group.
+
+        Returns
+        -------
+        W0 : ArrayLike of shape (n_samples_i, n_samples_i)
+            The inverse of the kernel matrix from the first group.
+        W1 : NDArraArrayLike of shape (n_samples_j, n_samples_j)
+            The inverse of the kernel matrix from the second group.
+
+        Notes
+        -----
+        Compute the W matrix for the estimated conditional average in
+        the KCD test :footcite:`Park2021conditional`.
+
+        References
+        ----------
+        .. footbibliography::
+        """
+        # compute kernel matrices
+        first_mask = np.array(1 - z, dtype=bool)
+        second_mask = np.array(z, dtype=bool)
+
+        # TODO: CHECK THAT THIS IS CORRECT
+        K0 = K[np.ix_(first_mask, first_mask)]
+        K1 = K[np.ix_(second_mask, second_mask)]
+
+        # compute the number of samples in each
+        n0 = int(np.sum(1 - z))
+        n1 = int(np.sum(z))
+
+        if isinstance(self.l2, (int, float)):
```

Comment:
It kind of seems like this logic should be packed in a property getter. `self.get_regs()`

Reply:
Will modify in #81